### ISSUE TITLE
Generated dataclass from CTE/WITH query should have nullable instance variable

### ISSUE THREAD

**Issue by maxstubbersfield** [simolus3/moor#1188]
Hi there,
Firstly, thanks heaps for maintaining such an awesome package! It is greatly appreciated :)

I have CTE as a custom query in a DAO and the dataclass that is generated uses a non-nullable data type for an instance variable which should actually nullable.

I have created a simple example to better illustrate the issue with two tables.  The `employees` table and another table with `employee_notes`. The WITH query gets the employee hierarchy - the employees that have no manager, then the employees that have those employees as managers and so on.
The result is the employee hierarchy and the notes on each employee. 

**tables.moor**
```
CREATE TABLE IF NOT EXISTS "employees" (
	"id"	INTEGER NOT NULL PRIMARY KEY,
	"name"	TEXT NOT NULL UNIQUE,
	"manager_id" INTEGER
);

CREATE TABLE IF NOT EXISTS "employee_notes" (
	"employee_id"	INTEGER NOT NULL PRIMARY KEY,
	"note"	TEXT
);

@create: INSERT INTO "employees" ("id", "name", "manager_id")
VALUES (1, 'Alice', NULL),
       (2, 'Bob', NULL),
       (3, 'Charlie', 1),
       (4, 'David', 2);

@create: INSERT INTO "employee_notes" ("employee_id", "note")
VALUES (3, 'A note about Charlie');
```
**employee.dao**
```
@UseDao(include: {
  'tables.moor'
}, tables: [
  Employee,
  EmployeeNote,
], queries: {
  'employeeHierarchy': '''
WITH RECURSIVE
employeeHierarchy(id, name, manager_id) AS (
  SELECT id, 
         name,
	 manager_id
    FROM employees
   WHERE manager_id IS NULL
   UNION ALL
  SELECT e.id, 
         e.name,
	 e.manager_id
    FROM employees e
    JOIN employeeHierarchy ON e.manager_id = employeeHierarchy.id
)
SELECT e.id, 
       e.name,
       e.manager_id,
       n.note
  FROM employeeHierarchy e
  LEFT OUTER JOIN
       employee_notes n ON e.id = n.employee_id
 ORDER BY e.id;
  '''
})
```
**Generated dataclass**
Here is where the issue is. The `managerId` should be `final int? managerId`, and as a result an exception is thrown when the query is parsed.
```
class EmployeeHierarchyResult {
  final int id;
  final String name;
  final int managerId;
  final String? note;
  EmployeeHierarchyResult({
    required this.id,
    required this.name,
    required this.managerId,
    this.note,
  });
}
```

I can work around this by doing the query and parsing the result manually but I thought I should log an issue anyway :)

Cheers!

**Comment by simolus3** [comment]
Thanks for the very helpful report! After some digging, it looks like this is actually caused by the compound select statement. When the type inference algorithm has to unify two types (which needs to happen to resolve the column in a `UNION`), it didn't look at the nullability of the involved columns. This should be fixed in 0d8741a7b456d6ce1cc50379421ec7e414c44a48.

**Comment by maxstubbersfield** [comment]
Thank you very much!!
